---
title: CNV enrichment patterns in epilepsy
author: Jean Monlong
output:
  md_document:
    variant: markdown_github
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo=TRUE, message=FALSE, warning=FALSE, fig.width=10)
```

# CNV enrichment patterns in epilepsy

## Load packages, functions and data

```{r}
library(knitr)
library(RColorBrewer)
library(PopSV)
library(dplyr)
library(magrittr)
library(ggplot2)
library(GenomicRanges)
library(parallel)
source("EpiPopSV-scripts.R")

## Load genomic annotations
load("../data/gencodev19-gviz-simp.RData")
gtf.simp %<>% mutate(chromosome=gsub("chr","",chromosome))
load("../data/SVdatabase.RData")
load("../data/genomicFeatures.RData")
## geneId -> geneName
gg = gen.feat.l$gene %>% select(geneId, geneName) %>% unique
geneidToGenename = gg$geneName
names(geneidToGenename) = gg$geneId
## Exons
exons.grl = list(exon=subset(gen.feat.l$exon, geneType=="protein_coding"))
## ExAC lof intolerant
exac = read.csv("../data/fordist_cleaned_exac_r03_march16_z_pli_rec_null_data.txt", as.is=TRUE, sep="\t")
gene.intolerant = subset(exac, pLI>.9)$gene
exons.grl$exon.pli = subset(gen.feat.l$exon, geneName %in% gene.intolerant)
## Epilepsy genes
epilepsy.genes = scan("../data/EpilepsyGenes.txt", "a", quiet=TRUE)
exons.grl$exon.epi = subset(gen.feat.l$exon, geneName %in% epilepsy.genes)
## Convert to GRanges
exons.grl = lapply(exons.grl, makeGRangesFromDataFrame, keep.extra.columns=TRUE)
## Enhancer map
load("../data/enhancerMap.RData")
eqtl.epi = subset(eqtl, gene_id %in% unique(exons.grl$exon.epi$geneId))
eqtl.epi$gene = geneidToGenename[eqtl.epi$gene_id]
seqlevels(dnaseMap) = gsub("chr","",seqlevels(dnaseMap))
dnaseMap.epi = subset(dnaseMap, gene %in% epilepsy.genes)
## Load and annotate CNVs
cnv.all = read.table("../data/cnvs-5kb_bins-EpiPopSV-allRefs.tsv.gz", header=TRUE, as.is=TRUE, sep="\t")
cnv.all = cnv.all %>% rmOL(1) %>% mutate(project=ifelse(project=="affected", "epilepsy", "parent"))
cnv.all.gr = makeGRangesFromDataFrame(cnv.all, keep.extra.columns=TRUE)
cnv.all$prop.db = dbProp(cnv.all.gr, svs.gr)
cnv.all$exon = overlapsAny(cnv.all.gr, exons.grl$exon)
cont.gr = reduce(subset(cnv.all.gr, project!="epilepsy"))
epi.gr = reduce(subset(cnv.all.gr, project=="epilepsy"))
cnv.all$parents = overlapsAny(cnv.all.gr, cont.gr)
cnv.all$exon.epi = overlapsAny(cnv.all.gr, exons.grl$exon.epi)
cnv.all$enhancer.epi = overlapsAny(cnv.all.gr, dnaseMap.epi) | overlapsAny(cnv.all.gr, eqtl.epi)
epid = distanceToNearest(cnv.all.gr, exons.grl$exon.epi) %>% as.data.frame
cnv.all$exon.epi.d = Inf
cnv.all$exon.epi.d[epid$queryHits] = epid$distance
cnv.all$exon.epi.closest = NA
cnv.all$exon.epi.closest[epid$queryHits] = exons.grl$exon.epi$geneName[epid$subjectHits]
cnv.all %<>% group_by(project) %>% do(freq.range(., annotate.only = TRUE)) %>% ungroup
```

Also let's choose how many cores we want to use:

```{r}
NB.CORES = 3
```

## Exonic enrichment

### Sub-sampling and permutation

NOTE: This is 20 times less sub-sampling and permutations than in the paper, hence the signal might not be as clean. It might take more time but feel free to increase `NB.SUBSAMP` and `NB.PERMS`.

```{r, cache=TRUE}
permF <-  function(rr, cnv.df, feat.grl){
    cat.all = cnv.df %>% group_by(project) %>% filter(sample %in% sample(unique(sample), 150,TRUE)) %>% do(reduceDf(.)) %>% mutate(sample=project, control=FALSE) %>% makeGRangesFromDataFrame(keep.extra.columns=TRUE)
    cat.null = draw.controls(cat.all, list(centel=gen.feat.l$centel), nb.cores=1)
    cat.null$project = cat.null$sample
    cat.null$sample = cat.all$sample = NULL
    cat.null$control = TRUE
    cat.all = c(cat.all, cat.null)
    res = lapply(1:length(feat.grl), function(ii){
                     cat.all$ol = overlapsAny(cat.all, feat.grl[[ii]])
                     obs = mcols(cat.all) %>% as.data.frame %>% group_by(control, project) %>% summarize(nb=sum(ol), prop=(1+nb)/(1+n())) %>% mutate(feat=names(feat.grl)[ii], rep=rr, set="obs")
                     exp = mcols(cat.all) %>% as.data.frame %>% mutate(project=sample(project)) %>% group_by(control, project) %>% summarize(nb=sum(ol), prop=(1+nb)/(1+n())) %>% mutate(feat=names(feat.grl)[ii], rep=rr, set="exp")
                     rbind(obs, exp)
                 })
    do.call(rbind, res)
}

NB.SUBSAMP = 10
NB.PERMS = 20
exons.all.pli = exons.grl
exons.all.pli$exon.epi = NULL
enr.all = do.call(rbind, mclapply(1:(NB.SUBSAMP*NB.PERMS), permF, cnv.df=cnv.all, feat.grl=exons.all.pli, mc.cores=NB.CORES))
enr.rare = do.call(rbind, mclapply(1:(NB.SUBSAMP*NB.PERMS), permF, cnv.df=subset(cnv.all, prop.db<.01), feat.grl=exons.all.pli, mc.cores=NB.CORES))
```

### Graphs for a batch of sub-sampling

```{r}
enr.all.ss = enr.all %>% filter(set=="obs", rep<=NB.SUBSAMP) %>% group_by(feat, project, rep) %>% summarize(enr=prop[!control]/prop[control])
ggplot(enr.all.ss, aes(x=reorder(feat, enr, mean), fill=project, y=enr)) + geom_boxplot() + theme_bw() + xlab("") + ylab("fold-enrichment") + coord_flip() + ggtitle("SV enrichment in exons") + scale_x_discrete(breaks=c("exon","exon.pli"), labels=c("all genes", "LoF\nintolerant\ngenes")) + theme(legend.position="bottom")

enr.rare.ss = enr.rare %>% filter(set=="obs", rep<=NB.SUBSAMP) %>% group_by(feat, project, rep) %>% summarize(enr=prop[!control]/prop[control])
ggplot(enr.rare.ss, aes(x=reorder(feat, enr, mean), fill=project, y=enr)) + geom_boxplot() + theme_bw() + xlab("") + ylab("fold-enrichment") + coord_flip() + ggtitle("Rare SV enrichment in exons") + scale_x_discrete(breaks=c("exon","exon.pli"), labels=c("all genes", "LoF\nintolerant\ngenes")) + theme(legend.position="bottom")
```

### Permutation test

It takes time to draw the control regions for each sub-sampling, hence the number of sub-sampling and permutation has been decreased in this report. For the paper high performance computing was used to run a thousand times a hundred sub-sampling.

```{r}
enr.all.ss.med = enr.all.ss %>% group_by(feat, project) %>% summarize(enr.med=median(enr))
enr.all %>% filter(set=="exp") %>% mutate(batch=cut(rep,NB.PERMS)) %>% group_by(feat, project, rep, batch) %>% summarize(enr=prop[!control]/prop[control]) %>% group_by(project, feat, batch) %>% summarize(enr=median(enr)) %>% merge(enr.all.ss.med) %>% group_by(feat, batch) %>% summarize(epi.enr=enr[project=="epilepsy"]-enr[project!="epilepsy"], epi.med=enr.med[project=="epilepsy"]-enr.med[project!="epilepsy"], epi.enr=epi.enr>epi.med) %>% group_by(feat) %>% summarize(pv=(1+sum(epi.enr))/(1+n())) %>% kable

enr.rare.ss.med = enr.rare.ss %>% group_by(feat, project) %>% summarize(enr.med=median(enr))
enr.rare %>% filter(set=="exp") %>% mutate(batch=cut(rep,NB.PERMS)) %>% group_by(feat, project, rep, batch) %>% summarize(enr=prop[!control]/prop[control]) %>% group_by(project, feat, batch) %>% summarize(enr=median(enr)) %>% merge(enr.rare.ss.med) %>% group_by(feat, batch) %>% summarize(epi.enr=enr[project=="epilepsy"]-enr[project!="epilepsy"], epi.med=enr.med[project=="epilepsy"]-enr.med[project!="epilepsy"], epi.enr=epi.enr>epi.med) %>% group_by(feat) %>% summarize(pv=(1+sum(epi.enr))/(1+n())) %>% kable
```

### Permutation test: a resampling trick

**This is not what was used for the paper**, but rather an approximation of the time-consuming permutations used for the paper.

We can approximate the permuted P-value by resampling the sub-sampling.

```{r}
NEW.NB.PERMS = 1000
NEW.NB.SUBSAMP = 30
bootstrapEnr <- function(df, nb.bs, nb.ss){
    data.frame(batch=1:nb.bs, enr=sapply(1:nb.bs, function(ii) median(sample(df$enr, nb.ss))))
}

enr.all %>% filter(set=="exp") %>% group_by(feat, project, rep) %>% summarize(enr=prop[!control]/prop[control]) %>% group_by(project, feat) %>% do(bootstrapEnr(., NEW.NB.PERMS, NEW.NB.SUBSAMP)) %>% merge(enr.all.ss.med) %>% group_by(feat, batch) %>% summarize(epi.enr=enr[project=="epilepsy"]-enr[project!="epilepsy"], epi.med=enr.med[project=="epilepsy"]-enr.med[project!="epilepsy"], epi.enr=epi.enr>epi.med) %>% group_by(feat) %>% summarize(pv=(1+sum(epi.enr))/(1+n())) %>% kable

enr.rare %>% filter(set=="exp") %>% group_by(feat, project, rep) %>% summarize(enr=prop[!control]/prop[control]) %>% group_by(project, feat) %>% do(bootstrapEnr(., NEW.NB.PERMS, NEW.NB.SUBSAMP)) %>% merge(enr.rare.ss.med) %>% group_by(feat, batch) %>% summarize(epi.enr=enr[project=="epilepsy"]-enr[project!="epilepsy"], epi.med=enr.med[project=="epilepsy"]-enr.med[project!="epilepsy"], epi.enr=epi.enr>epi.med) %>% group_by(feat) %>% summarize(pv=(1+sum(epi.enr))/(1+n())) %>% kable
```


## Rare exonic CNVs are less private in the epilepsy cohort

### All epilepsy patients, controls down-sampled

```{r}
rare.ex = cnv.all %>% filter(prop.db<.01, exon)
rare.ex.f = rare.ex %>% group_by(project) %>% do(freqSS(.))
rare.ex.f %>% group_by(project, nb, rep) %>% summarize(n=n()/nb[1]) %>% group_by(project, rep) %>% arrange(desc(nb)) %>% mutate(cn=cumsum(n), cprop=cn/sum(n)) %>% filter(nb>1, nb<11) %>% group_by(project, nb) %>% summarize(cprop.5=quantile(cprop, probs=.05), cprop.95=quantile(cprop, probs=.95), cprop=median(cprop)) %>% ggplot(aes(x=nb, y=cprop, colour=project, fill=project)) + geom_ribbon(aes(ymin=cprop.5, ymax=cprop.95), colour=FALSE, alpha=.2) + geom_line(alpha=.7) + geom_point(size=2) + scale_colour_brewer(name="", palette="Set1") + scale_fill_brewer(name="", palette="Set1") + theme_bw() + scale_x_continuous(breaks=2:10, labels=paste0(c(2:10),"+")) + xlab("CNV recurrence") + ylab("proportion of rare exonic CNVs")
```

### Same after exteme samples removal

In each cohort, let's remove the top 20 samples with the most rare exonic non-private CNVs.

```{r}
rare.epi.samp.o = rare.ex.f %>% filter(nb>1) %>% group_by(project, sample) %>% summarize(cnv=n()) %>% arrange(desc(cnv)) %>% group_by(project) %>% do(head(.,20))
rare.ex.f.noext = rare.ex %>% filter(!(sample %in% rare.epi.samp.o$sample)) %>% group_by(project) %>% do(freqSS(., sample.size=178))
rare.ex.f.noext %>% group_by(project, nb, rep) %>% summarize(n=n()/nb[1]) %>% group_by(project, rep) %>% arrange(desc(nb)) %>% mutate(cn=cumsum(n), cprop=cn/sum(n)) %>% filter(nb>1, nb<11) %>% group_by(project, nb) %>% summarize(cprop.5=quantile(cprop, probs=.05), cprop.95=quantile(cprop, probs=.95), cprop=median(cprop)) %>% ggplot(aes(x=nb, y=cprop, colour=project, fill=project)) + geom_ribbon(aes(ymin=cprop.5, ymax=cprop.95), colour=FALSE, alpha=.2) + geom_line(alpha=.7) + geom_point(size=2) + scale_colour_brewer(name="", palette="Set1") + scale_fill_brewer(name="", palette="Set1") + theme_bw() + scale_x_continuous(breaks=2:10, labels=paste0(c(2:10),"+")) + xlab("CNV recurrence") + ylab("proportion of rare exonic CNVs") + ggtitle("Top 20 most extreme samples removed")
```

### French-Canadians only

```{r}
french.canadians = scan("../data/frenchCanadiansControls.txt", "a", quiet=TRUE)
cnv.fc = cnv.all %>% filter(project=="epilepsy" | sample %in% french.canadians)
fc.ss = cnv.fc %>% select(sample, project) %>% unique %>% group_by(project) %>% summarize(n=n()) %>% .$n %>% unlist %>% min
rare.ex.fc = cnv.fc %>% filter(exon, prop.db<.01) %>% group_by(project) %>% do(freqSS(., sample.size=fc.ss))
rare.ex.fc %>% group_by(project, nb, rep) %>% summarize(n=n()/nb[1]) %>% group_by(project, rep) %>% arrange(desc(nb)) %>% mutate(cn=cumsum(n), cprop=cn/sum(n)) %>% filter(nb>1, nb<11) %>% group_by(project, nb) %>% summarize(cprop.5=quantile(cprop, probs=.05), cprop.95=quantile(cprop, probs=.95), cprop=median(cprop)) %>% ggplot(aes(x=nb, y=cprop, colour=project, fill=project)) + geom_ribbon(aes(ymin=cprop.5, ymax=cprop.95), colour=FALSE, alpha=.2) + geom_line(alpha=.7) + geom_point(size=2) + scale_colour_brewer(name="", palette="Set1") + scale_fill_brewer(name="", palette="Set1") + theme_bw() + scale_x_continuous(breaks=2:10, labels=paste0(c(2:10),"+")) + xlab("CNV recurrence") + ylab("proportion of rare exonic CNVs") + ggtitle("French-Canadians only")
```

### Permutation test

```{r, cache=TRUE}
diff.obs = rare.ex.f %>% group_by(project, nb) %>% summarize(n=n()/nb[1]) %>% group_by(project) %>% summarize(non.priv=sum(n[nb>1])/sum(n)) %>% arrange(desc(project)) %>% .$non.priv %>% diff
diff.exp = replicate(1000, {
  rare.ex.f %>% ungroup %>% mutate(project=sample(project)) %>% group_by(project, nb) %>% summarize(n=n()/nb[1]) %>% group_by(project) %>% summarize(non.priv=sum(n[nb>1])/sum(n)) %>% arrange(desc(project)) %>% .$non.priv %>% diff
})

diff.obs.noext = rare.ex.f.noext %>% group_by(project, nb) %>% summarize(n=n()/nb[1]) %>% group_by(project) %>% summarize(non.priv=sum(n[nb>1])/sum(n)) %>% arrange(desc(project)) %>% .$non.priv %>% diff
diff.exp.noext = replicate(1000, {
  rare.ex.f.noext %>% ungroup %>% mutate(project=sample(project)) %>% group_by(project, nb) %>% summarize(n=n()/nb[1]) %>% group_by(project) %>% summarize(non.priv=sum(n[nb>1])/sum(n)) %>% arrange(desc(project)) %>% .$non.priv %>% diff
})

diff.obs.fc = rare.ex.fc %>% group_by(project, nb) %>% summarize(n=n()/nb[1]) %>% group_by(project) %>% summarize(non.priv=sum(n[nb>1])/sum(n)) %>% arrange(desc(project)) %>% .$non.priv %>% diff
diff.exp.fc = replicate(1000, {
  rare.ex.fc %>% ungroup %>% mutate(project=sample(project)) %>% group_by(project, nb) %>% summarize(n=n()/nb[1]) %>% group_by(project) %>% summarize(non.priv=sum(n[nb>1])/sum(n)) %>% arrange(desc(project)) %>% .$non.priv %>% diff
})
```

```{r}
priv.test = data.frame(test=c("all samples", "top 20 extreme samples removed", "french-canadian"),
           pv=c((1+sum(diff.obs<=diff.exp))/(1+length(diff.exp)),
               (1+sum(diff.obs.noext<=diff.exp.noext))/(1+length(diff.exp.noext)),
                       (1+sum(diff.obs.fc<=diff.exp.fc))/(1+length(diff.exp.fc))))
kable(priv.test)
```

## Non-coding rare CNVs

```{r, cache=TRUE}
freqEpiRareSS <- function(df, sample.size=198, nb.rep=100){
    if(length(unique(df$sample)) > sample.size){
        res = mclapply(1:nb.rep, function(ii){
                           df %>% filter(sample %in% sample(unique(sample),sample.size)) %>% filter(exon.epi.d<3e6) %>% do(freq.range(.,annotate.only=TRUE)) %>% filter(exon.epi.d<4e5, prop.db<.01) %>% mutate(rep=ii) %>% as.data.frame
                       }, mc.cores=NB.CORES)
        return(do.call(rbind, res))
    } else {
        return(df %>% freq.range(annotate.only=TRUE) %>% filter(exon.epi.d<4e5, prop.db<.01) %>% mutate(rep=1))
    }
}
cnv.ss = cnv.all %>% group_by(project) %>% do(freqEpiRareSS(., nb.rep=100))
```

### All rare non-coding CNVs

```{r}
cnb.nc = cnv.ss %>% filter(prop.db<.01, nb<10, !exon) %>% group_by(project, rep) %>% arrange(exon.epi.d) %>% filter(!duplicated(sample)) %>% do(decdf(., "exon.epi.d")) %>% group_by(project, d) %>% summarize(cnb.5=quantile(cnb, probs=.05), cnb.95=quantile(cnb, probs=.95), cnb=median(cnb))
ggplot(subset(cnb.nc, d<5e4), aes(x=d/1e3, y=cnb, colour=project, fill=project)) + geom_line(size=2) + geom_ribbon(aes(ymax=cnb.95, ymin=cnb.5), alpha=.2, linetype=2, data=subset(cnb.nc, d<5e4 & project=="parent")) + theme_bw() + xlab("distance to nearest epilepsy exon (kb)") + ylab("cumulative affected samples")
```

### Rare non-coding CNVs that overlaps functional annotations

Focusing on CNVs that overlaps an eQTL for the epilepsy gene, or a DNase I hypersensitive site associated to the promoter of the epilepsy gene.

```{r}
cnb.df = cnv.ss %>% filter(prop.db<.01, nb<10, !exon, enhancer.epi) %>% group_by(project, rep) %>% arrange(exon.epi.d) %>% filter(!duplicated(sample)) %>% do(decdf(., "exon.epi.d")) %>% group_by(project, d) %>% summarize(cnb.5=quantile(cnb, probs=.05), cnb.95=quantile(cnb, probs=.95), cnb=median(cnb))
ggplot(cnb.df, aes(x=d/1e3, y=cnb, colour=project, fill=project)) + geom_line(size=2) + geom_ribbon(aes(ymax=cnb.95, ymin=cnb.5), alpha=.2, linetype=2, data=subset(cnb.df, project=="parent")) + theme_bw() + xlab("distance to nearest epilepsy exon (kb)") + ylab("cumulative affected samples")

cnb.rec = cnv.ss %>% filter(prop.db<.01, nb>1, nb<10, !exon, enhancer.epi) %>% group_by(project, rep) %>% arrange(exon.epi.d) %>% filter(!duplicated(sample)) %>% do(decdf(., "exon.epi.d")) %>% group_by(project, d) %>% summarize(cnb.5=quantile(cnb, probs=.05), cnb.95=quantile(cnb, probs=.95), cnb=median(cnb))
ggplot(cnb.rec, aes(x=d/1e3, y=cnb, colour=project, fill=project)) + geom_line(size=2) + geom_ribbon(aes(ymax=cnb.95, ymin=cnb.5), alpha=.2, linetype=2, data=subset(cnb.rec, project=="parent")) + theme_bw() + xlab("distance to nearest epilepsy exon (kb)") + ylab("cumulative affected samples") + ggtitle("Non-private")
```


## Epilepsy-specific deletions hitting epilepsy genes

```{r}
epi.dels = cnv.all %>% filter(!parents, z<0)
```

### Fisher test

```{r}
exon.gr = exons.grl$exon
exon.cnv.genes = epi.dels %>% makeGRangesFromDataFrame(keep.extra.columns=TRUE) %>% subsetByOverlaps(exon.gr, .) %>% mcols %>% .$geneName %>% unique
ft.table = table(unique(exon.gr$geneName) %in% epilepsy.genes)
ft.table = rbind(ft.table, table(exon.cnv.genes %in% epilepsy.genes))
fisher.test(ft.table)
```

### Controlling for the gene size

Epilepsy genes tend to be large, and larger genes are more likely hit by variants (including CNVs). We can select genes with similar gene size in order to control for the size effect. We check that in the distribution of gene size are satisfactory.

```{r}
gene.sum = exon.gr %>% as.data.frame %>% group_by(geneName) %>% summarize(exon.size=sum(end-start), start=min(start), end=max(end), size=end-start, nb.exon=n()) %>% mutate(epilepsy=geneName %in% epilepsy.genes, cnv=geneName %in% exon.cnv.genes)
size.bk = c(quantile(subset(gene.sum, cnv)$size, probs=seq(0,1,.2)), Inf)
gene.sum$size.class = cut(gene.sum$size, size.bk, include.lowest = TRUE)
gene.sum.cnv = subset(gene.sum, cnv)
gene.cont.ii = lapply(levels(gene.sum.cnv$size.class), function(sc){
                          sample(which(gene.sum$size.class==sc), sum(gene.sum.cnv$size.class==sc))
                      })
gene.cont = gene.sum$geneName[unlist(gene.cont.ii)]
gene.sum.n = rbind(gene.sum %>% mutate(set="all genes"),
                   gene.sum %>% filter(geneName %in% gene.cont) %>% mutate(set="CNV-hit genes control"),
                   gene.sum %>% filter(epilepsy) %>% mutate(set="epilepsy genes"),
                   gene.sum %>% filter(cnv) %>% mutate(set="CNV-hit genes"))

colpal = brewer.pal(3,"Set1")[c(1,2,2,3)]
ggplot(gene.sum.n, aes(x=size, colour=set, linetype=set)) + stat_density(size=2, geom="line", position="dodge") + theme_bw() + scale_x_log10() + scale_linetype_manual(values=c(1,1,2,1), name="") + scale_colour_manual(values=colpal, name="") + ylab("density") + xlab("gene size (bp)") + theme(legend.position=c(0,1), legend.justification=c(0,1))
```

Eventially we can check that the exonic sequence size and number of exons are also controlled by this approach.

```{r}
ggplot(gene.sum.n, aes(x=exon.size, colour=set, linetype=set)) + stat_density(size=2, geom="line", position="dodge") + theme_bw() + scale_x_log10() + scale_linetype_manual(values=c(1,1,2,1), name="") + scale_colour_manual(values=colpal, name="") + ylab("density") + xlab("exonic size (bp)") + theme(legend.position=c(0,1), legend.justification=c(0,1))
ggplot(gene.sum.n, aes(x=nb.exon, colour=set, linetype=set)) + stat_density(size=2, geom="line", position="dodge") + theme_bw() + scale_x_log10() + scale_linetype_manual(values=c(1,1,2,1), name="") + scale_colour_manual(values=colpal, name="") + ylab("density") + xlab("number of exons") + theme(legend.position=c(0,1), legend.justification=c(0,1))
```

Then the new sampling-based test:

```{r}
testGenesCnvs <- function(cnv.df, nb.perm=1e3, nb.class=5){
  exon.cnv.genes = cnv.df %>% makeGRangesFromDataFrame(keep.extra.columns=TRUE) %>% subsetByOverlaps(exon.gr, .) %>% mcols %>% .$geneName %>% unique
  gene.sum %<>%  mutate(cnv=geneName %in% exon.cnv.genes)
  size.bk = c(quantile(subset(gene.sum, cnv)$size, probs=seq(0,1,1/nb.class)), Inf)
  gene.sum$size.class = cut(gene.sum$size, size.bk, include.lowest = TRUE)
  gene.sum.cnv = subset(gene.sum, cnv)
  obs.epi = sum(exon.cnv.genes %in% epilepsy.genes)
  exp.epi.size = mclapply(1:nb.perm, function(ii){
    gene.cont.ii = lapply(levels(gene.sum.cnv$size.class), function(sc){
      sample(which(gene.sum$size.class==sc), sum(gene.sum.cnv$size.class==sc))
    })
    gene.cont = gene.sum$geneName[unlist(gene.cont.ii)]
    sum(gene.cont %in% epilepsy.genes)
  }, mc.cores=NB.CORES)
  data.frame(odds.ratio = obs.epi/mean(unlist(exp.epi.size)),
             gene=nrow(gene.sum.cnv),
             gene.epi=obs.epi,
             gene.epi.cont=mean(unlist(exp.epi.size)),
             pv=(sum(obs.epi<=unlist(exp.epi.size))+1)/(length(exp.epi.size)+1))
}

test.epi.genes.sum = rbind(
    testGenesCnvs(epi.dels, nb.perm=2e3, nb.class=1) %>% mutate(test="all genes, deletions not in parents"),
    testGenesCnvs(epi.dels, nb.perm=1e4, nb.class=5) %>% mutate(test="size-controlled genes, deletions not in parents"),
    testGenesCnvs(subset(epi.dels, prop.db==0), nb.perm=1e4, nb.class=5) %>% mutate(test="size-controlled genes, deletions not in parents or controls")
    )

kable(test.epi.genes.sum)
```
