---
title: CNV enrichment patterns in epilepsy
author: Jean Monlong
output:
  md_document:
    variant: markdown_github
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo=TRUE, message=FALSE, warning=FALSE, fig.width=10, tidy=TRUE)
set.seed(666)
```

# CNV enrichment patterns in epilepsy

## Load packages, functions and data

```{r}
library(knitr)
library(RColorBrewer)
library(PopSV)
library(dplyr)
library(magrittr)
library(ggplot2)
library(GenomicRanges)
library(parallel)
library(tidyr)
source("EpiPopSV-scripts.R")

## Load genomic annotations
load("../data/SVdatabase.RData")
load("../data/genomicFeatures.RData")
## geneId -> geneName
gg = gen.feat.l$gene %>% select(geneId, geneName) %>% unique
geneidToGenename = gg$geneName
names(geneidToGenename) = gg$geneId
## Exons
exons.grl = list(exon=subset(gen.feat.l$exon, geneType=="protein_coding"))
## ExAC lof intolerant
exac = read.csv("../data/fordist_cleaned_exac_r03_march16_z_pli_rec_null_data.txt", as.is=TRUE, sep="\t")
gene.intolerant = subset(exac, pLI>.9)$gene
exons.grl$exon.pli = subset(gen.feat.l$exon, geneName %in% gene.intolerant)
## Epilepsy genes
epilepsy.genes = scan("../data/EpilepsyGenes.txt", "a", quiet=TRUE)
exons.grl$exon.epi = subset(gen.feat.l$exon, geneName %in% epilepsy.genes)
## Convert to GRanges
exons.grl = lapply(exons.grl, makeGRangesFromDataFrame, keep.extra.columns=TRUE)
## Enhancer map
load("../data/enhancerMap.RData")
eqtl.epi = subset(eqtl, gene_id %in% unique(exons.grl$exon.epi$geneId))
eqtl.epi$gene = geneidToGenename[eqtl.epi$gene_id]
seqlevels(dnaseMap) = gsub("chr","",seqlevels(dnaseMap))
dnaseMap.epi = subset(dnaseMap, gene %in% epilepsy.genes)
## Load and annotate CNVs
cnv.all = read.table("../data/cnvs-PopSV-Epilepsy-198affected-301controls-5kb.tsv.gz", header=TRUE, as.is=TRUE, sep="\t")
cnv.all = cnv.all %>% mutate(project=ifelse(project=="affected", "patients", "controls"))
cnv.all.gr = makeGRangesFromDataFrame(cnv.all, keep.extra.columns=TRUE)
cnv.all$prop.db = dbProp(cnv.all.gr, svs.gr)
cnv.all$prop.db.50 = cnv.all.gr %>% dbProp(., svs.gr, min.db.span=.5) # 50% overlap
cnv.all$exon = overlapsAny(cnv.all.gr, exons.grl$exon)
epi.gr = reduce(subset(cnv.all.gr, project=="patients"))
cnv.all$exon.epi = overlapsAny(cnv.all.gr, exons.grl$exon.epi)
cnv.all$enhancer.epi = overlapsAny(cnv.all.gr, dnaseMap.epi) | overlapsAny(cnv.all.gr, eqtl.epi)
epid = distanceToNearest(cnv.all.gr, exons.grl$exon.epi) %>% as.data.frame
cnv.all$exon.epi.d = Inf
cnv.all$exon.epi.d[epid$queryHits] = epid$distance
cnv.all$exon.epi.closest = NA
cnv.all$exon.epi.closest[epid$queryHits] = exons.grl$exon.epi$geneName[epid$subjectHits]
cnv.all %<>% group_by(project) %>% do(freq.range(., annotate.only = TRUE)) %>% ungroup
info.df = cnv.all %>% select(sample, project) %>% unique
```

Also let's choose how many cores we want to use:

```{r}
NB.CORES = 3
```

## Exonic enrichment

This part was run on a high-performance computing cluster.
The code can be found in `epilepsy-enrichmentPatterns-HPC.R`.

Briefly, 150 samples were sub-sampled 100 times in the patient and in the control cohorts. 
For each sub-sampling, the CNVs are merged per cohort and the catalogs are overlapped with exonic sequences.
All genes and genes with a predicted loss-of-function intolerance (pLi>0.9) were used.

To control for the CNV size distribution, control regions were randomly selected in the reference genome controlling for the size and overlap with assembly gaps.
To test for differences between patients and controls, CNVs from the two cohorts were permuted 10,000 times and we saved the difference between the medians of the two groups (see boxplots below).

We performed the analysis separately on *large* CNVs (>=50 Kbp) and *small* CNVs (<50 Kbp).
We also run the analysis using rare CNVs (<1% frequency in the external SV databases) only.

```{r}
load('../data/permExCnv.RData')

enr.obs = lapply(permExCnv, function(l)l$enr.obs)
enr.obs = do.call(rbind, enr.obs)

enr.obs %<>% mutate(freq=factor(freq, levels=c('all','rare'), labels=c("all CNVs","rare CNVs")), size=factor(size, levels=c('large', 'small'), labels=c("> 50 Kbp","< 50 Kbp")))

ggplot(enr.obs, aes(x=feat, y=enr, fill=project)) + geom_boxplot(notch=TRUE) + theme_bw() + facet_grid(size~freq, scales='free') + xlab("") + ylab("fold-enrichment") + scale_x_discrete(breaks=c("exon","exon.pli"), labels=c("all genes", "LoF\nintolerant\ngenes")) + geom_hline(yintercept=1, linetype=2) + scale_fill_brewer(name="",palette="Set1") + theme(legend.position=c(.01,.01), legend.justification=c(0,0))

enr.exp.diff = lapply(permExCnv, function(l)l$enr.exp.diff)
enr.exp.diff = do.call(rbind, enr.exp.diff)
enr.exp.diff %<>% mutate(set=factor(paste(freq, size), levels=c('all large', 'rare large', 'all small', 'rare small'), labels=c('all CNVs > 50 Kbp', 'rare CNVs > 50 Kbp', 'all CNVs < 50 Kbp', 'rare CNVs < 50 Kbp')))

enr.obs.diff = enr.obs %>% mutate(set=factor(paste(freq, size), levels=c('all CNVs > 50 Kbp', 'rare CNVs > 50 Kbp', 'all CNVs < 50 Kbp', 'rare CNVs < 50 Kbp'))) %>% group_by(set, feat, project) %>% summarize(enr=median(enr)) %>% group_by(set, feat) %>% summarize(diff=enr[project=='patients']-enr[project=='controls'])

ggplot(enr.exp.diff, aes(x=feat, y=diff))  + geom_hline(yintercept=0, linetype=2) + geom_violin(fill='grey90') + geom_point(data=enr.obs.diff, size=5, color='red') + theme_bw() + facet_wrap(~set, scales='free', ncol=2) + xlab("") + ylab("difference between patients and controls") + scale_x_discrete(breaks=c("exon","exon.pli"), labels=c("all genes", "LoF\nintolerant\ngenes")) + scale_fill_brewer(name="",palette="Set1")
```

*The grey violin plot shows the distribution of the differences across the 10,000 permutations. The red dot represents the difference observed between the patients and controls.*

The empirical P-value from the 10,000 permutations:

```{r}
enr.obs.diff %>% dplyr::rename(diff.obs=diff) %>% merge(enr.exp.diff) %>% group_by(set, feat) %>% summarize(pv.enr=(sum(diff>=diff.obs)+1)/(n()+1), pv.dep=(sum(diff<=diff.obs)+1)/(n()+1)) %>% kable
```


## Rare exonic CNVs are less private in the epilepsy cohort

### All epilepsy patients, controls down-sampled

```{r}
rare.ex = cnv.all %>% filter(prop.db<.01, exon)
rare.ex.f = rare.ex %>% group_by(project) %>% do(freqSS(.))
rare.ex.f %>% group_by(project, nb, rep) %>% summarize(n=n()/nb[1]) %>% group_by(project, rep) %>% arrange(desc(nb)) %>% mutate(cn=cumsum(n), cprop=cn/sum(n)) %>% filter(nb>1, nb<11) %>% group_by(project, nb) %>% summarize(cprop.5=quantile(cprop, probs=.05), cprop.95=quantile(cprop, probs=.95), cprop=median(cprop)) %>% ggplot(aes(x=nb, y=cprop, colour=project, fill=project)) + geom_ribbon(aes(ymin=cprop.5, ymax=cprop.95), colour=FALSE, alpha=.2) + geom_line(alpha=.7) + geom_point(size=2) + scale_colour_brewer(name="", palette="Set1") + scale_fill_brewer(name="", palette="Set1") + theme_bw() + scale_x_continuous(breaks=2:10, labels=paste0(c(2:10),"+")) + xlab("CNV recurrence") + ylab("proportion of rare exonic CNVs") + theme(legend.position=c(.99,.99), legend.justification=c(1,1))
```

### Same after exteme samples removal

In each cohort, let's remove the top 20 samples with the most rare exonic non-private CNVs.

```{r}
rare.epi.samp.o = rare.ex.f %>% filter(nb>1) %>% group_by(project, sample) %>% summarize(cnv=n()) %>% arrange(desc(cnv)) %>% group_by(project) %>% do(head(.,20))
rare.ex.f.noext = rare.ex %>% filter(!(sample %in% rare.epi.samp.o$sample)) %>% group_by(project) %>% do(freqSS(., sample.size=178))
rare.ex.f.noext %>% group_by(project, nb, rep) %>% summarize(n=n()/nb[1]) %>% group_by(project, rep) %>% arrange(desc(nb)) %>% mutate(cn=cumsum(n), cprop=cn/sum(n)) %>% filter(nb>1, nb<11) %>% group_by(project, nb) %>% summarize(cprop.5=quantile(cprop, probs=.05), cprop.95=quantile(cprop, probs=.95), cprop=median(cprop)) %>% ggplot(aes(x=nb, y=cprop, colour=project, fill=project)) + geom_ribbon(aes(ymin=cprop.5, ymax=cprop.95), colour=FALSE, alpha=.2) + geom_line(alpha=.7) + geom_point(size=2) + scale_colour_brewer(name="", palette="Set1") + scale_fill_brewer(name="", palette="Set1") + theme_bw() + scale_x_continuous(breaks=2:10, labels=paste0(c(2:10),"+")) + xlab("CNV recurrence") + ylab("proportion of rare exonic CNVs") + ggtitle("Top 20 most extreme samples removed") + theme(legend.position=c(.99,.99), legend.justification=c(1,1))
```

### French-Canadians only

```{r}
french.canadians = scan("../data/frenchCanadiansControls.txt", "a", quiet=TRUE)
cnv.fc = cnv.all %>% filter(project=="patients" | sample %in% french.canadians)
fc.ss = cnv.fc %>% select(sample, project) %>% unique %>% group_by(project) %>% summarize(n=n()) %>% .$n %>% unlist %>% min
rare.ex.fc = cnv.fc %>% filter(exon, prop.db<.01) %>% group_by(project) %>% do(freqSS(., sample.size=fc.ss))
rare.ex.fc %>% group_by(project, nb, rep) %>% summarize(n=n()/nb[1]) %>% group_by(project, rep) %>% arrange(desc(nb)) %>% mutate(cn=cumsum(n), cprop=cn/sum(n)) %>% filter(nb>1, nb<11) %>% group_by(project, nb) %>% summarize(cprop.5=quantile(cprop, probs=.05), cprop.95=quantile(cprop, probs=.95), cprop=median(cprop)) %>% ggplot(aes(x=nb, y=cprop, colour=project, fill=project)) + geom_ribbon(aes(ymin=cprop.5, ymax=cprop.95), colour=FALSE, alpha=.2) + geom_line(alpha=.7) + geom_point(size=2) + scale_colour_brewer(name="", palette="Set1") + scale_fill_brewer(name="", palette="Set1") + theme_bw() + scale_x_continuous(breaks=2:10, labels=paste0(c(2:10),"+")) + xlab("CNV recurrence") + ylab("proportion of rare exonic CNVs") + ggtitle("French-Canadians only") + theme(legend.position=c(.99,.99), legend.justification=c(1,1))
```

### Permutation test

```{r, cache=TRUE}
diff.obs = rare.ex.f %>% group_by(project, nb) %>% summarize(n=n()/nb[1]) %>% group_by(project) %>% summarize(non.priv=sum(n[nb>1])/sum(n)) %>% arrange(project) %>% .$non.priv %>% diff
diff.exp = replicate(1000, {
  rare.ex.f %>% ungroup %>% mutate(project=sample(project)) %>% group_by(project, nb) %>% summarize(n=n()/nb[1]) %>% group_by(project) %>% summarize(non.priv=sum(n[nb>1])/sum(n)) %>% arrange(project) %>% .$non.priv %>% diff
})

diff.obs.noext = rare.ex.f.noext %>% group_by(project, nb) %>% summarize(n=n()/nb[1]) %>% group_by(project) %>% summarize(non.priv=sum(n[nb>1])/sum(n)) %>% arrange(project) %>% .$non.priv %>% diff
diff.exp.noext = replicate(1000, {
  rare.ex.f.noext %>% ungroup %>% mutate(project=sample(project)) %>% group_by(project, nb) %>% summarize(n=n()/nb[1]) %>% group_by(project) %>% summarize(non.priv=sum(n[nb>1])/sum(n)) %>% arrange(project) %>% .$non.priv %>% diff
})

diff.obs.fc = rare.ex.fc %>% group_by(project, nb) %>% summarize(n=n()/nb[1]) %>% group_by(project) %>% summarize(non.priv=sum(n[nb>1])/sum(n)) %>% arrange(project) %>% .$non.priv %>% diff
diff.exp.fc = replicate(1000, {
  rare.ex.fc %>% ungroup %>% mutate(project=sample(project)) %>% group_by(project, nb) %>% summarize(n=n()/nb[1]) %>% group_by(project) %>% summarize(non.priv=sum(n[nb>1])/sum(n)) %>% arrange(project) %>% .$non.priv %>% diff
})
```

```{r}
priv.test = data.frame(test=c("all samples", "top 20 extreme samples removed", "french-canadian"),
           pv=c((1+sum(diff.obs<=diff.exp))/(1+length(diff.exp)),
               (1+sum(diff.obs.noext<=diff.exp.noext))/(1+length(diff.exp.noext)),
                       (1+sum(diff.obs.fc<=diff.exp.fc))/(1+length(diff.exp.fc))))
kable(priv.test)
```

## Non-coding rare CNVs

```{r, cache=TRUE}
freqEpiRareSS <- function(df, sample.size=198, nb.rep=100){
    if(length(unique(df$sample)) > sample.size){
        res = mclapply(1:nb.rep, function(ii){
                           df %>% filter(sample %in% sample(unique(sample),sample.size)) %>% filter(exon.epi.d<3e6) %>% do(freq.range(.,annotate.only=TRUE)) %>% filter(exon.epi.d<4e5, prop.db<.01) %>% mutate(rep=ii) %>% as.data.frame
                       }, mc.cores=NB.CORES)
        return(do.call(rbind, res))
    } else {
        return(df %>% freq.range(annotate.only=TRUE) %>% filter(exon.epi.d<4e5, prop.db<.01) %>% mutate(rep=1))
    }
}
cnv.ss = cnv.all %>% group_by(project) %>% do(freqEpiRareSS(., nb.rep=100))
```

### All rare non-coding CNVs

```{r}
cnb.nc = cnv.ss %>% filter(prop.db<.01, nb<10, !exon) %>% group_by(project, rep) %>% arrange(exon.epi.d) %>% filter(!duplicated(sample)) %>% filter(exon.epi.d<3e5)
cnb.nc %>% group_by(project, rep) %>% do(decdf(., "exon.epi.d")) %>% group_by(project, d) %>% summarize(cnb.5=quantile(cnb, probs=.05), cnb.95=quantile(cnb, probs=.95), cnb=median(cnb))  %>% filter(d<3e5) %>% ggplot(aes(x=d/1e3, y=cnb, colour=project, fill=project)) + geom_line(size=2) + geom_ribbon(aes(ymax=cnb.95, ymin=cnb.5), linetype=2, alpha=.2) + theme_bw() + xlab("distance to nearest epilepsy exon (kb)") + ylab("cumulative affected samples") + scale_fill_brewer(name="",palette="Set1") + scale_colour_brewer(name="",palette="Set1") + guides(alpha=FALSE) + theme(legend.position=c(.99,.01), legend.justification=c(1,0))

cnb.nc.or = cnb.nc %>% group_by(project, rep) %>% do(decdf(., "exon.epi.d")) %>% group_by(project, d) %>% summarize(cnb=median(cnb)) %>% filter(d<3e5) %>% spread(project, cnb) %>% mutate(odds.ratio=(patients/(198-patients))/(controls/(198-controls)))
ggplot(cnb.nc.or, aes(x=d/1e3, y=log(odds.ratio))) + geom_area(alpha=.1) + geom_line() + theme_bw() + xlab("distance to nearest epilepsy exon (kb)") + ylab("log odds ratio") + geom_hline(yintercept=0, linetype=2)
cnb.nc.or %>% filter(d %in% c(5e3, 5e4, 1e5)) %>% kable

ks.test(cnb.nc$exon.epi.d[which(cnb.nc$project=='controls')], cnb.nc$exon.epi.d[which(cnb.nc$project!='controls')])
wilcox.test(cnb.nc$exon.epi.d[which(cnb.nc$project=='controls')], cnb.nc$exon.epi.d[which(cnb.nc$project!='controls')])
```

#### Deletions

```{r ncdel}
cnb.nc.del = cnv.ss %>% filter(z<0, prop.db<.01, nb<10, !exon) %>% group_by(project, rep) %>% arrange(exon.epi.d) %>% filter(!duplicated(sample)) %>% filter(exon.epi.d<3e5)
cnb.nc.del %>% group_by(project, rep) %>% do(decdf(., "exon.epi.d")) %>% group_by(project, d) %>% summarize(cnb.5=quantile(cnb, probs=.05), cnb.95=quantile(cnb, probs=.95), cnb=median(cnb))  %>% filter(d<3e5) %>% ggplot(aes(x=d/1e3, y=cnb, colour=project, fill=project)) + geom_line(size=2) + geom_ribbon(aes(ymax=cnb.95, ymin=cnb.5), linetype=2, alpha=.2) + theme_bw() + xlab("distance to nearest epilepsy exon (kb)") + ylab("cumulative affected samples") + scale_fill_brewer(name="",palette="Set1") + scale_colour_brewer(name="",palette="Set1") + guides(alpha=FALSE) + theme(legend.position=c(.99,.01), legend.justification=c(1,0))

cnb.nc.del.or = cnb.nc.del %>% group_by(project, rep) %>% do(decdf(., "exon.epi.d")) %>% group_by(project, d) %>% summarize(cnb=median(cnb)) %>% filter(d<3e5) %>% spread(project, cnb) %>% mutate(odds.ratio=(patients/(198-patients))/(controls/(198-controls)))
ggplot(cnb.nc.del.or, aes(x=d/1e3, y=log(odds.ratio))) + geom_area(alpha=.1) + geom_line() + theme_bw() + xlab("distance to nearest epilepsy exon (kb)") + ylab("log odds ratio") + geom_hline(yintercept=0, linetype=2)
cnb.nc.del.or %>% filter(d %in% c(5e3, 5e4, 1e5)) %>% kable

ks.test(cnb.nc.del$exon.epi.d[which(cnb.nc.del$project=='controls')], cnb.nc.del$exon.epi.d[which(cnb.nc.del$project!='controls')])
wilcox.test(cnb.nc.del$exon.epi.d[which(cnb.nc.del$project=='controls')], cnb.nc.del$exon.epi.d[which(cnb.nc.del$project!='controls')])
```

To go further we test each epilepsy gene to see if it has more rare non-coding deletions in the epilepsy cohort versus the controls.

```{r ncdelgene}
cnv.all %>% filter(!exon, prop.db<.01, exon.epi.d<2e5, z<0) %>% group_by(exon.epi.closest, project) %>% summarize(sample=length(unique(sample))) %>% spread(project, sample, fill=0) %>% mutate(pv=fisher.test(cbind(c(301-controls, controls), c(198-patients, patients)))$p.value) %>% arrange(pv) %>% ungroup %>% head(10) %>% kable
```

#### Duplications

```{r ncdup}
cnb.nc.dup = cnv.ss %>% filter(z>0, prop.db<.01, nb<10, !exon) %>% group_by(project, rep) %>% arrange(exon.epi.d) %>% filter(!duplicated(sample)) %>% filter(exon.epi.d<3e5)
cnb.nc.dup %>% group_by(project, rep) %>% do(decdf(., "exon.epi.d")) %>% group_by(project, d) %>% summarize(cnb.5=quantile(cnb, probs=.05), cnb.95=quantile(cnb, probs=.95), cnb=median(cnb))  %>% filter(d<3e5) %>% ggplot(aes(x=d/1e3, y=cnb, colour=project, fill=project)) + geom_line(size=2) + geom_ribbon(aes(ymax=cnb.95, ymin=cnb.5), linetype=2, alpha=.2) + theme_bw() + xlab("distance to nearest epilepsy exon (kb)") + ylab("cumulative affected samples") + scale_fill_brewer(name="",palette="Set1") + scale_colour_brewer(name="",palette="Set1") + guides(alpha=FALSE) + theme(legend.position=c(.99,.01), legend.justification=c(1,0))

cnb.nc.dup.or = cnb.nc.dup %>% group_by(project, rep) %>% do(decdf(., "exon.epi.d")) %>% group_by(project, d) %>% summarize(cnb=median(cnb)) %>% filter(d<3e5) %>% spread(project, cnb) %>% mutate(odds.ratio=(patients/(198-patients))/(controls/(198-controls)))
ggplot(cnb.nc.dup.or, aes(x=d/1e3, y=log(odds.ratio))) + geom_area(alpha=.1) + geom_line() + theme_bw() + xlab("distance to nearest epilepsy exon (kb)") + ylab("log odds ratio") + geom_hline(yintercept=0, linetype=2)
cnb.nc.dup.or %>% filter(d %in% c(5e3, 5e4, 1e5)) %>% kable

ks.test(cnb.nc.dup$exon.epi.d[which(cnb.nc.dup$project=='controls')], cnb.nc.dup$exon.epi.d[which(cnb.nc.dup$project!='controls')])
wilcox.test(cnb.nc.dup$exon.epi.d[which(cnb.nc.dup$project=='controls')], cnb.nc.dup$exon.epi.d[which(cnb.nc.dup$project!='controls')])
```



### Rare non-coding CNVs that overlaps functional annotations

Focusing on CNVs that overlaps an eQTL for the epilepsy gene, or a DNase I hypersensitive site associated to the promoter of the epilepsy gene.

```{r}
cnb.df = cnv.ss %>% filter(prop.db<.01, nb<10, !exon, enhancer.epi) %>% group_by(project, rep) %>% arrange(exon.epi.d) %>% filter(!duplicated(sample))  %>% filter(exon.epi.d<3e5)
cnb.df %>% group_by(project, rep) %>% do(decdf(., "exon.epi.d")) %>% group_by(project, d) %>% summarize(cnb.5=quantile(cnb, probs=.05), cnb.95=quantile(cnb, probs=.95), cnb=median(cnb))  %>% filter(d<3e5) %>% ggplot(aes(x=d/1e3, y=cnb, colour=project, fill=project)) + geom_line(size=2) + geom_ribbon(aes(ymax=cnb.95, ymin=cnb.5), alpha=.2, linetype=2) + theme_bw() + xlab("distance to nearest epilepsy exon (kb)") + ylab("cumulative affected samples") + scale_fill_brewer(name="",palette="Set1") + scale_colour_brewer(name="",palette="Set1") + theme(legend.position=c(.99,.01), legend.justification=c(1,0))

cnb.or = cnb.df %>% group_by(project, rep) %>% do(decdf(., "exon.epi.d")) %>% group_by(project, d) %>% summarize(cnb=median(cnb)) %>% filter(d<3e5) %>% spread(project, cnb) %>% mutate(odds.ratio=(patients/(198-patients))/(controls/(198-controls)))
ggplot(cnb.or, aes(x=d/1e3, y=log(odds.ratio))) + geom_area(alpha=.1) + geom_line() + theme_bw() + xlab("distance to nearest epilepsy exon (kb)") + ylab("log odds ratio") + geom_hline(yintercept=0, linetype=2)
cnb.or %>% filter(d %in% c(5e3, 5e4, 1e5)) %>% kable

ks.test(cnb.df$exon.epi.d[which(cnb.df$project=='controls')], cnb.df$exon.epi.d[which(cnb.df$project!='controls')])
wilcox.test(cnb.df$exon.epi.d[which(cnb.df$project=='controls')], cnb.df$exon.epi.d[which(cnb.df$project!='controls')])

cnb.rec = cnv.ss %>% filter(prop.db<.01, nb>1, nb<10, !exon, enhancer.epi) %>% group_by(project, rep) %>% arrange(exon.epi.d) %>% filter(!duplicated(sample)) %>% filter(exon.epi.d<3e5)
cnb.rec %>% do(decdf(., "exon.epi.d")) %>% group_by(project, d) %>% summarize(cnb.5=quantile(cnb, probs=.05), cnb.95=quantile(cnb, probs=.95), cnb=median(cnb)) %>% filter(d<3e5) %>% ggplot(aes(x=d/1e3, y=cnb, colour=project, fill=project)) + geom_line(size=2) + geom_ribbon(aes(ymax=cnb.95, ymin=cnb.5), alpha=.2, linetype=2) + theme_bw() + xlab("distance to nearest epilepsy exon (kb)") + ylab("cumulative affected samples") + ggtitle("Non-private") + scale_fill_brewer(name="",palette="Set1") + scale_colour_brewer(name="",palette="Set1") + theme(legend.position=c(.99,.01), legend.justification=c(1,0))

cnb.rec.or = cnb.rec %>% group_by(project, rep) %>% do(decdf(., "exon.epi.d")) %>% group_by(project, d) %>% summarize(cnb=median(cnb)) %>% filter(d<3e5) %>% spread(project, cnb) %>% mutate(odds.ratio=(patients/(198-patients))/(controls/(198-controls)))
ggplot(cnb.rec.or, aes(x=d/1e3, y=log(odds.ratio))) + geom_area(alpha=.1) + geom_line() + theme_bw() + xlab("distance to nearest epilepsy exon (kb)") + ylab("log odds ratio") + geom_hline(yintercept=0, linetype=2)
cnb.rec.or %>% filter(d %in% c(5e3, 5e4, 1e5)) %>% kable

ks.test(cnb.rec$exon.epi.d[which(cnb.rec$project=='controls')], cnb.rec$exon.epi.d[which(cnb.rec$project!='controls')])
wilcox.test(cnb.rec$exon.epi.d[which(cnb.rec$project=='controls')], cnb.rec$exon.epi.d[which(cnb.rec$project!='controls')])
```

#### Deletions

```{r ncdelenh}
cnb.enh.del = cnv.ss %>% filter(z<0, prop.db<.01, nb<10, !exon, enhancer.epi) %>% group_by(project, rep) %>% arrange(exon.epi.d) %>% filter(!duplicated(sample)) %>% filter(exon.epi.d<3e5)
cnb.enh.del %>% group_by(project, rep) %>% do(decdf(., "exon.epi.d")) %>% group_by(project, d) %>% summarize(cnb.5=quantile(cnb, probs=.05), cnb.95=quantile(cnb, probs=.95), cnb=median(cnb))  %>% filter(d<3e5) %>% ggplot(aes(x=d/1e3, y=cnb, colour=project, fill=project)) + geom_line(size=2) + geom_ribbon(aes(ymax=cnb.95, ymin=cnb.5), linetype=2, alpha=.2) + theme_bw() + xlab("distance to nearest epilepsy exon (kb)") + ylab("cumulative affected samples") + scale_fill_brewer(name="",palette="Set1") + scale_colour_brewer(name="",palette="Set1") + guides(alpha=FALSE) + theme(legend.position=c(.99,.01), legend.justification=c(1,0))
```

#### Duplications

```{r ncdupenh}
cnb.enh.dup = cnv.ss %>% filter(z>0, prop.db<.01, nb<10, !exon, enhancer.epi) %>% group_by(project, rep) %>% arrange(exon.epi.d) %>% filter(!duplicated(sample)) %>% filter(exon.epi.d<3e5)
cnb.enh.dup %>% group_by(project, rep) %>% do(decdf(., "exon.epi.d")) %>% group_by(project, d) %>% summarize(cnb.5=quantile(cnb, probs=.05), cnb.95=quantile(cnb, probs=.95), cnb=median(cnb))  %>% filter(d<3e5) %>% ggplot(aes(x=d/1e3, y=cnb, colour=project, fill=project)) + geom_line(size=2) + geom_ribbon(aes(ymax=cnb.95, ymin=cnb.5), linetype=2, alpha=.2) + theme_bw() + xlab("distance to nearest epilepsy exon (kb)") + ylab("cumulative affected samples") + scale_fill_brewer(name="",palette="Set1") + scale_colour_brewer(name="",palette="Set1") + guides(alpha=FALSE) + theme(legend.position=c(.99,.01), legend.justification=c(1,0))
```



## Rare CNVs hitting epilepsy genes

### Controlling for the gene size

Epilepsy genes tend to be large, and larger genes are more likely hit by variants (including CNVs). We can select genes with similar gene size in order to control for the size effect. We check that in the distribution of gene size are satisfactory.

```{r}
exon.cnv.genes = cnv.all %>% filter(project=="patients") %>% makeGRangesFromDataFrame(keep.extra.columns=TRUE) %>% subsetByOverlaps(exons.grl$exon, .) %>% mcols %>% .$geneName %>% unique
gene.sum = exons.grl$exon %>% as.data.frame %>% group_by(geneName) %>% summarize(exon.size=sum(end-start), start=min(start), end=max(end), size=end-start, nb.exon=n()) %>% mutate(epilepsy=geneName %in% epilepsy.genes, cnv=geneName %in% exon.cnv.genes)
size.bk = c(quantile(subset(gene.sum, cnv)$size, probs=seq(0,1,.2)), Inf)
gene.sum$size.class = cut(gene.sum$size, size.bk, include.lowest = TRUE)
gene.sum.cnv = subset(gene.sum, cnv)
gene.cont.ii = lapply(levels(gene.sum.cnv$size.class), function(sc){
                          sample(which(gene.sum$size.class==sc), sum(gene.sum.cnv$size.class==sc))
                      })
gene.cont = gene.sum$geneName[unlist(gene.cont.ii)]
gene.sum.n = rbind(gene.sum %>% mutate(set="all genes"),
                   gene.sum %>% filter(geneName %in% gene.cont) %>% mutate(set="CNV-hit genes control"),
                   gene.sum %>% filter(epilepsy) %>% mutate(set="epilepsy genes"),
                   gene.sum %>% filter(cnv) %>% mutate(set="CNV-hit genes"))

colpal = brewer.pal(3,"Set1")[c(1,2,2,3)]
ggplot(gene.sum.n, aes(x=size, colour=set, linetype=set)) + stat_density(size=2, geom="line", position="dodge") + theme_bw() + scale_x_log10() + scale_linetype_manual(values=c(1,1,2,1), name="") + scale_colour_manual(values=colpal, name="") + ylab("density") + xlab("gene size (bp)") + theme(legend.position=c(0,1), legend.justification=c(0,1))
```

Eventially we can check that the exonic sequence size and number of exons are also controlled by this approach.

```{r}
ggplot(gene.sum.n, aes(x=exon.size, colour=set, linetype=set)) + stat_density(size=2, geom="line", position="dodge") + theme_bw() + scale_x_log10() + scale_linetype_manual(values=c(1,1,2,1), name="") + scale_colour_manual(values=colpal, name="") + ylab("density") + xlab("exonic size (bp)") + theme(legend.position=c(0,1), legend.justification=c(0,1))
ggplot(gene.sum.n, aes(x=nb.exon, colour=set, linetype=set)) + stat_density(size=2, geom="line", position="dodge") + theme_bw() + scale_x_log10() + scale_linetype_manual(values=c(1,1,2,1), name="") + scale_colour_manual(values=colpal, name="") + ylab("density") + xlab("number of exons") + theme(legend.position=c(0,1), legend.justification=c(0,1))
```

Then the new sampling-based test:

```{r}
testGenesCnvs <- function(cnv.df, nb.perm=1e3, nb.class=5){
  exon.cnv.genes = cnv.df %>% makeGRangesFromDataFrame(keep.extra.columns=TRUE) %>% subsetByOverlaps(exons.grl$exon, .) %>% mcols %>% .$geneName %>% unique
  gene.sum %<>%  mutate(cnv=geneName %in% exon.cnv.genes)
  size.bk = c(quantile(subset(gene.sum, cnv)$size, probs=seq(0,1,1/nb.class)), Inf)
  gene.sum$size.class = cut(gene.sum$size, size.bk, include.lowest = TRUE)
  gene.sum.cnv = subset(gene.sum, cnv)
  obs.epi = sum(exon.cnv.genes %in% epilepsy.genes)
  exp.epi.size = mclapply(1:nb.perm, function(ii){
    gene.cont.ii = lapply(levels(gene.sum.cnv$size.class), function(sc){
      sample(which(gene.sum$size.class==sc), sum(gene.sum.cnv$size.class==sc))
    })
    gene.cont = gene.sum$geneName[unlist(gene.cont.ii)]
    sum(gene.cont %in% epilepsy.genes)
  }, mc.cores=NB.CORES)
  list(exp=unlist(exp.epi.size), obs=obs.epi,
    sum.df = data.frame(fold.enr = obs.epi/mean(unlist(exp.epi.size)),
                        gene=nrow(gene.sum.cnv),
                        gene.epi=obs.epi,
                        gene.epi.cont=mean(unlist(exp.epi.size)),
                        pv=(sum(obs.epi<=unlist(exp.epi.size))+1)/(length(exp.epi.size)+1))
  )
}
```

### Exonic deletions absent from the public databases

```{r}
test.epi.genes = list(
  allGenesDelNoDB=testGenesCnvs(cnv.all %>% filter(project=="patients", z<0, prop.db==0), nb.perm=1e3, nb.class=1),
  sizeGenesDelNoDB=testGenesCnvs(cnv.all %>% filter(project=="patients", z<0, prop.db==0), nb.perm=1e3, nb.class=5)
)

test.epi.genes.sum = do.call(rbind, lapply(names(test.epi.genes), function(x) data.frame(test=x, test.epi.genes[[x]]$sum.df)))
kable(test.epi.genes.sum)

test.df = rbind(data.frame(exp=test.epi.genes$allGenesDelNoDB$exp, test="all genes"),
                data.frame(exp=test.epi.genes$sizeGenesDelNoDB$exp, test="size-controlled genes"))
ggplot(test.df, aes(x=exp, fill=test)) + geom_histogram(position="dodge", binwidth=1) + theme_bw() + geom_vline(xintercept=test.epi.genes$sizeGenesDelNoDB$obs, linetype=2) + xlab("number of epilepsy genes among sampled genes") + ylab("number of sampling") + scale_fill_brewer(palette="Set2",name="sampling") + theme(legend.position="bottom") + ggtitle("Genes hit deletions never seen in public databases") + annotate("text", x=test.epi.genes$sizeGenesDelNoDB$obs, y=100, label="observed", vjust=-1, angle=-90)
```

### Exploring the enrichments

Here, the number of permutations has been decreased compared to what was used in the study.

```{r}
cnv.sets = cnv.all %>% mutate(type=ifelse(z<0, "deletion","duplication")) %>% select(project, sample, chr, start, end, type) %>% as.data.frame
cnv.sets$prop.db = cnv.sets %>% makeGRangesFromDataFrame(keep.extra.columns=TRUE) %>% dbProp(., svs.gr)

load("../data/cnvs-PopSV-twin-5kbp-FDR001.RData")
tw.sets = res.df %>% mutate(project="Twins", type=ifelse(z<0, "deletion","duplication")) %>% select(project, sample, chr, start, end, type)
tw.sets$prop.db = tw.sets %>% makeGRangesFromDataFrame(keep.extra.columns=TRUE) %>% dbProp(., subset(svs.gr, project!="PopSV"))
cnv.sets = rbind(cnv.sets, tw.sets)
```

```{r, cache=TRUE}
freq.r = c(1,.1,.01,.001,.0001, 0) 
enr.epi.freq = lapply(freq.r, function(freq.max){
  cnv.sets %>% filter(prop.db <= freq.max) %>% group_by(project, type) %>% do(testGenesCnvs(., nb.perm=1e3)$sum.df) %>% mutate(freq=freq.max)
})
enr.epi.freq = do.call(rbind, enr.epi.freq)
```

```{r}
ggplot(enr.epi.freq, aes(x=factor(freq), y=fold.enr, colour=project, group=project)) + geom_line() + geom_point(aes(size=cut(pv, c(0,.05,1)))) + theme_bw() + ylim(0, max(enr.epi.freq$fold.enr)) + geom_hline(yintercept=1, linetype=3) + scale_size_manual(name="P-value", labels=c("<0.05",">0.05"), values=c(4,2)) + xlab("CNV frequency in public databases") + facet_grid(type~.) + ylab("fold-enrichment") + scale_colour_brewer(name="",palette="Set1")
```

```{r, include=FALSE}
save(enr.obs, enr.exp.diff, enr.obs.diff, rare.ex.f, rare.ex.f.noext, rare.ex.fc, cnb.nc, cnb.nc.del, cnb.df, cnb.or, gene.sum.n, test.df, test.epi.genes, enr.epi.freq, file='epilepsy-enrichmentPatterns.RData')
```

